.github/workflows/tests.yml

```diff
@@ -356,28 +391,92 @@ jobs:
       fromJson(needs.setup_tests.outputs.test-dependents)
+    runs-on: ubuntu-22.04
+    container:
+      image: ghcr.io/homebrew/ubuntu22.04:master
+    outputs:
+      runners: ${{steps.determine-dependent-runners.outputs.runners}}
+      runners_present: ${{steps.determine-dependent-runners.outputs.runners_present}}
+    env:
+      HOMEBREW_LINUX_RUNNER: ${{needs.setup_tests.outputs.linux-runner}}
+      HOMEBREW_LINUX_CLEANUP: ${{needs.setup_tests.outputs.cleanup-linux-runner}}
+      TESTING_FORMULAE: ${{needs.formulae_detect.outputs.testing_formulae}}
+    steps:
+      - name: Set up Homebrew
+        id: set-up-homebrew
+        uses: Homebrew/actions/setup-homebrew@master
+
+      - name: Determine runners to use for test_deps job
+        id: determine-dependent-runners
+        run: brew determine-test-runners --dependents --eval-all "$TESTING_FORMULAE"
+
+  conclusion:
+    needs: [tests, setup_tests, setup_runners]
+    if: always() && github.event_name == 'pull_request'
+    runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: /bin/bash -e {0}
+    steps:
+      - name: Check `tests` result
+        run: |
+          result='${{ needs.tests.result }}'
+          printf '::notice ::`tests` job status: %s\n' "$result"
+
+          # Possible values are `success`, `failure`, `cancelled` or `skipped`.
+          # https://docs.github.com/en/actions/learn-github-actions/contexts#needs-context
+          if [[ "$result" = "failure" ]] || [[ "$result" = "cancelled" ]]
+          then
+            printf '::error ::`tests` job %s.\n' "$result"
+            exit 1
+          fi
+
+          runners_present='${{ needs.setup_runners.outputs.runners_present }}'
+          syntax_only='${{ needs.setup_tests.outputs.syntax-only }}'
+
+          # The tests job can be skipped only if the PR is syntax-only
+          # or no runners were assigned.
+          if [[ "$result" = "skipped" ]] &&
+             [[ "$runners_present" = "false" || "$syntax_only" = "true" ]]
+          then
+            exit 0
+          fi
+
+          # The test job can succeed only if the PR is not syntax-only
+          # and runners were assigned. Otherwise it must have been skipped.
+          if [[ "$result" = "success" ]] &&
+             [[ "$runners_present" = "true" ]] &&
+             [[ "$syntax_only" = "false" ]]
+          then
+            exit 0
+          fi
+
+          # If we made it here, something went wrong with our workflow run that needs investigating.
+          printf '::error ::Unexpected outcome!\n'
+          printf '::error ::`tests` job result: %s\n' "$result" # success/skipped
+          printf '::error ::runners assigned:   %s\n' "$runners_present" # true/false
+          printf '::error ::syntax-only:        %s\n' "$syntax_only" # true/false
+          exit 1
+
+  test_deps:
+    needs: [tap_syntax, setup_tests, setup_dep_runners, tests]
+    if: >
+      (success() || failure()) &&
```

Comment:
I might get some pushback here: but this just restores the original behaviour of testing dependents before we split it into a separate job. (i.e. it is run even when the `tests` job fails)

Reply:
@carlocab Yeh, I'd rather we didn't do this please. It wastes runner time on jobs we're not going to ship.

Reply:
It really is still needed for some PRs. As a compromise: I've made it so that `test_deps` is run despite failing `tests` only when https://github.com/Homebrew/homebrew-core/labels/CI-no-fail-fast is set.

Reply:
> It really is still needed for some PRs.

Which ones and why? These are the sort of jobs that are going to prevent our ability to make `tests` a required CI status.


> As a compromise: I've made it so that test_deps is run despite failing tests only when [CI-no-fail-fast](https://github.com/Homebrew/homebrew-core/labels/CI-no-fail-fast) Continue CI tests despite failing GitHub Actions matrix builds. is set.

Works for now, happy to merge as-is here 👍🏻

Reply:
> Which ones and why?

Typically: audit failures or `GITHUB_TOKEN` expiry errors. The most common type of audit failure we end up ignoring is an online check failure (e.g. `brew audit` reports that the source tarball is not available but `brew install` and `brew fetch` seemed to have no problem with it -- bizarre, but it happens). Less commonly, re-running CI just takes too long (several hours) that it's just a better use of resources to fix the audit failure in a follow-up PR.

Reply:
> e.g. brew audit reports that the source tarball is not available

IMO we should just remove all these checks considering they are basically duplicates.

> GITHUB_TOKEN expiry errors

I think the new jobs should address this? Alternatively, using a BrewTestBot PAT?

Reply:
Alternatively: we just pull literally all the `brew audit` runs into a `tap_syntax` like job that's Linux-only, run on the detected formulae and run before and `tests` jobs even run. Easily rerun in that case, too.

Reply:
> IMO we should just remove all these checks considering they are basically duplicates.

Yea, 👍 on that. The only difficult part is that we have _a lot_ of audits, so it's a bit painful to go through them all to weed out the ones we don't want anymore.

> I think the new jobs should address this? Alternatively, using a BrewTestBot PAT?

New jobs will help. BrewTestBot PAT will not -- no secrets available on PR workflows.

> Alternatively: we just pull literally all the `brew audit` runs into a `tap_syntax` like job that's Linux-only, run on the detected formulae and run before and `tests` jobs even run. Easily rerun in that case, too.

I've been thinking this too. This might be the nicest solution.